### PR TITLE
Add hint about data proxy to invalid uri error [DPGA]

### PR DESCRIPTION
**Pull request by pimeys**
Part of: https://github.com/prisma/schema-team/issues/348

### CODE REVIEW

File: introspection-engine/introspection-engine-tests/tests/errors/mod.rs [hunk 1]
```diff
@@ -81,9 +81,9 @@ async fn bad_connection_string_in_datamodel_returns_nice_error() {
...
+            "message": "\u{1b}[1;91merror\u{1b}[0m: \u{1b}[1mError validating datasource `db`: the URL must start with the protocol `postgresql://` or `postgres://`.\n\nTo use a URL with protocal `prisma://` the Data Proxy must be enabled via `prisma generate --data-proxy`.\n\nMore information about Data Proxy: https://pris.ly/d/data-proxy\n\u{1b}[0m\n  \u{1b}[1;94m-->\u{1b}[0m  \u{1b}[4mschema.prisma:4\u{1b}[0m\n\u{1b}[1;94m   | \u{1b}[0m\n\u{1b}[1;94m 3 | \u{1b}[0m        provider = \"postgresql\"\n\u{1b}[1;94m 4 | \u{1b}[0m        url      = \u{1b}[1;91m\"sqlserver:/localhost:1433;database=prisma-demo;user=SA;password=Pr1sm4_Pr1sm4;trustServerCertificate=true;encrypt=true\"\u{1b}[0m\n\u{1b}[1;94m   | \u{1b}[0m\n",
```
Comment (janpio) [inline introspection-engine/introspection-engine-tests/tests/errors/mod.rs:84]:
```suggestion
            "message": "\u{1b}[1;91merror\u{1b}[0m: \u{1b}[1mError validating datasource `db`: the URL must start with the protocol `postgresql://` or `postgres://`.\n\nTo use a URL with protocol `prisma://` the Data Proxy must be enabled via `prisma generate --data-proxy`.\n\nMore information about Data Proxy: https://pris.ly/d/data-proxy\n\u{1b}[0m\n  \u{1b}[1;94m-->\u{1b}[0m  \u{1b}[4mschema.prisma:4\u{1b}[0m\n\u{1b}[1;94m   | \u{1b}[0m\n\u{1b}[1;94m 3 | \u{1b}[0m        provider = \"postgresql\"\n\u{1b}[1;94m 4 | \u{1b}[0m        url      = \u{1b}[1;91m\"sqlserver:/localhost:1433;database=prisma-demo;user=SA;password=Pr1sm4_Pr1sm4;trustServerCertificate=true;encrypt=true\"\u{1b}[0m\n\u{1b}[1;94m   | \u{1b}[0m\n",
```

File: introspection-engine/introspection-engine-tests/tests/errors/mod.rs [hunk 2]
```diff
@@ -81,9 +81,9 @@ async fn bad_connection_string_in_datamodel_returns_nice_error() {
...
+                "full_error": "\u{1b}[1;91merror\u{1b}[0m: \u{1b}[1mError validating datasource `db`: the URL must start with the protocol `postgresql://` or `postgres://`.\n\nTo use a URL with protocal `prisma://` the Data Proxy must be enabled via `prisma generate --data-proxy`.\n\nMore information about Data Proxy: https://pris.ly/d/data-proxy\n\u{1b}[0m\n  \u{1b}[1;94m-->\u{1b}[0m  \u{1b}[4mschema.prisma:4\u{1b}[0m\n\u{1b}[1;94m   | \u{1b}[0m\n\u{1b}[1;94m 3 | \u{1b}[0m        provider = \"postgresql\"\n\u{1b}[1;94m 4 | \u{1b}[0m        url      = \u{1b}[1;91m\"sqlserver:/localhost:1433;database=prisma-demo;user=SA;password=Pr1sm4_Pr1sm4;trustServerCertificate=true;encrypt=true\"\u{1b}[0m\n\u{1b}[1;94m   | \u{1b}[0m\n",
```
Comment (janpio) [inline introspection-engine/introspection-engine-tests/tests/errors/mod.rs:86]:
```suggestion
                "full_error": "\u{1b}[1;91merror\u{1b}[0m: \u{1b}[1mError validating datasource `db`: the URL must start with the protocol `postgresql://` or `postgres://`.\n\nTo use a URL with protocol `prisma://` the Data Proxy must be enabled via `prisma generate --data-proxy`.\n\nMore information about Data Proxy: https://pris.ly/d/data-proxy\n\u{1b}[0m\n  \u{1b}[1;94m-->\u{1b}[0m  \u{1b}[4mschema.prisma:4\u{1b}[0m\n\u{1b}[1;94m   | \u{1b}[0m\n\u{1b}[1;94m 3 | \u{1b}[0m        provider = \"postgresql\"\n\u{1b}[1;94m 4 | \u{1b}[0m        url      = \u{1b}[1;91m\"sqlserver:/localhost:1433;database=prisma-demo;user=SA;password=Pr1sm4_Pr1sm4;trustServerCertificate=true;encrypt=true\"\u{1b}[0m\n\u{1b}[1;94m   | \u{1b}[0m\n",
```

File: libs/datamodel/core/src/configuration/datasource.rs [hunk 3]
```diff
@@ -90,6 +90,14 @@ impl Datasource {
         };
 
         self.active_connector.validate_url(&url).map_err(|err_str| {
+            let err_str = indoc::formatdoc! {"
+                {err_str}
+
+                To use a URL with protocal `prisma://` the Data Proxy must be enabled via `prisma generate --data-proxy`.
```
Comment (janpio) [inline libs/datamodel/core/src/configuration/datasource.rs:96]:
```suggestion
                To use a URL with protocol `prisma://` the Data Proxy must be enabled via `prisma generate --data-proxy`.
```

File: libs/datamodel/core/tests/config/sources.rs [hunk 4]
```diff
@@ -284,7 +289,12 @@ fn must_error_if_wrong_protocol_is_used_for_postgresql() {
         .unwrap_err();
 
     let expectation = expect![[r#"
-        [1;91merror[0m: [1mError validating datasource `myds`: the URL must start with the protocol `postgresql://` or `postgres://`.[0m
+        [1;91merror[0m: [1mError validating datasource `myds`: the URL must start with the protocol `postgresql://` or `postgres://`.
+
+        To use a URL with protocal `prisma://` the Data Proxy must be enabled via `prisma generate --data-proxy`.
```
Comment (janpio) [inline libs/datamodel/core/tests/config/sources.rs:294]:
```suggestion
        To use a URL with protocol `prisma://` the Data Proxy must be enabled via `prisma generate --data-proxy`.
```

File: libs/datamodel/core/tests/config/sources.rs [hunk 5]
```diff
@@ -341,7 +351,12 @@ fn must_error_if_wrong_protocol_is_used_for_sqlite() {
         .unwrap_err();
 
     let expectation = expect![[r#"
-        [1;91merror[0m: [1mError validating datasource `myds`: the URL must start with the protocol `file:`.[0m
+        [1;91merror[0m: [1mError validating datasource `myds`: the URL must start with the protocol `file:`.
+
+        To use a URL with protocal `prisma://` the Data Proxy must be enabled via `prisma generate --data-proxy`.
```
Comment (janpio) [inline libs/datamodel/core/tests/config/sources.rs:356]:
```suggestion
        To use a URL with protocol `prisma://` the Data Proxy must be enabled via `prisma generate --data-proxy`.
```

**Comment by janpio** [comment]
Currently this is appended to all connection string protocol mismatch errors. 
It would be optimal if we could only append that when `prisma://` is actually used.
(Can be done in a follow up PR of course.)

**Review by janpio — APPROVED** [review]

**Comment by pimeys** [comment]
image

**Pull request merged**